mediorum/server/upload_client.go

```diff
@@ -0,0 +1,78 @@
+package server
+
+import (
+	"fmt"
+	"time"
+
+	"gorm.io/gorm/clause"
+)
+
+func (ss *MediorumServer) startUploadScroller() {
+	time.Sleep(time.Minute)
+
+	for {
+		for _, peer := range ss.Config.Peers {
+			if peer.Host == ss.Config.Self.Host {
+				continue
+			}
+
+			uploadCursor := &UploadCursor{
+				Host: peer.Host,
+			}
+
+			// load prior cursor for host
+			ss.crud.DB.First(&uploadCursor, "host = ?", peer.Host)
+			logger := ss.logger.With("task", "upload_scroll", "host", peer.Host, "after", uploadCursor.After)
+
+			// fetch uploads from host
+			var uploads []*Upload
+			u := apiPath(peer.Host, "uploads") + "?after=" + uploadCursor.After.Format(time.RFC3339Nano)
+
+			resp, err := ss.reqClient.R().
+				SetSuccessResult(&uploads).
+				Get(u)
+
+			if err != nil {
+				logger.Error("list uploads failed", "err", err)
+				continue
+			}
+			if resp.StatusCode != 200 {
+				err := fmt.Errorf("%s: %s %s", resp.Request.RawURL, resp.Status, string(resp.Bytes()))
+				logger.Error("list uploads failed", "err", err)
+				continue
+			}
+
+			var overwrites []*Upload
+			for _, upload := range uploads {
+				// get existing upload
+				var existing *Upload
+				ss.crud.DB.First(&existing, "id = ?", upload.ID)
+
+				overwrite := existing == nil || (existing.Status != JobStatusDone && upload.Status == JobStatusDone) || existing.TranscodedAt.Before(upload.TranscodedAt)
```

Comment:
for track previews, what if you change the middle condition to `|| (existing.Status != JobStatusDone && existing.Status != JobStatusRetranscode && upload.Status == JobStatusDone)`?

Reply:
also could be wrong but i think if no record is found then gorm would set `existing` to a struct with nil fields rather than making the `existing` pointer nil itself. it seems safer to check for a non-existent upload by doing `if ss.crud.DB.First(&existing, "id = ?", upload.ID).Error == gorm.ErrRecordNotFound {`